File: docs/palindromes-in-peg/scaffold_gs_heads.py
"""Finite GS head table -> one-instruction scaffold transitions.

The standalone fixture consumes a/b/# to load a snapshot, ! to finish it,
and . for local work. It is an instruction/compiler test, NOT a PAL grammar.
Snapshot loading and rewinding are real transitions, never injected pointers.
"""
from gs_heads import HEADS, BLIND, TESTS, compile_controller, unit_moves
from scaffold_circuit import (Circuit, Value, Ref, PREVIOUS, TRUE, FALSE,
                              conjunction, disjunction, neg, choose)
from scaffold_circuit_structs import Tape, StackPool
from scaffold_head_distances import HeadDistances
from scaffold_live_distances import LiveDistances


class GSHeads:
  def __init__(self, circuit, program=None, prefix="gs", compact=True):
    self.circuit = circuit
    self.program = unit_moves(compile_controller() if program is None else program)
    self.compact = compact
    self.bank = LiveDistances(circuit, self.program, HEADS, prefix=prefix + ".distance") if compact else \
                HeadDistances(circuit, HEADS, prefix=prefix + ".distance", exclusive_moves=True)
    self.pool = StackPool(circuit, {"cells": 1}, ("_", "a", "b", "#"))
    self.tapes = {head: Tape(circuit, prefix + "." + head, self.pool.payload, pool=self.pool)
                  for head in HEADS if head not in BLIND}
    self.pc_key, self.mode_key = prefix + ".pc", prefix + ".mode"
    self.pc = circuit.get(PREVIOUS, self.pc_key, tuple(range(len(self.program.code))), self.program.start)
    self.mode = circuit.get(PREVIOUS, self.mode_key, ("load", "rewind", "run", "done"), "load")
    self.found_key = prefix + ".found"
    self.found = circuit.get(PREVIOUS, self.found_key, (False, True), False).eq(True)
    self.border = FALSE

  def copy(self, target, source, enabled):
    if not self.compact:
      self.bank.copy(target, source, enabled)
    if target in self.tapes:
      if source not in self.tapes:
        raise ValueError("a data head cannot copy a blind coordinate")
      a, b = self.tapes[target], self.tapes[source]
      a.left.copy_from(b.left, enabled)
      a.right.copy_from(b.right, enabled)
      a.write(b.focus, enabled)

  def tick(self):
    circuit, char = self.circuit, self.circuit.input()
    load = conjunction(self.mode.eq("load"), disjunction(*(char.eq(s) for s in "ab#")))
    finish_load = conjunction(self.mode.eq("load"), char.eq("!"))
    rewind = conjunction(self.mode.eq("rewind"), char.eq("."))
    at_origin = self.tapes["Origin"].left.empty() if self.compact else self.bank.equal("Origin", "First")
    begin = conjunction(rewind, at_origin)
    active = conjunction(self.mode.eq("run"), char.eq("."))
    moves = {(head, d): [] for head in HEADS for d in (-1, 1)}
    copies, next_pc, halts, borders = {}, {}, [], []

    def edge(destination, enabled):
      next_pc[destination] = disjunction(next_pc.get(destination, FALSE), enabled)

    events = [conjunction(active, self.pc.eq(state)) for state in range(len(self.program.code))]
    for enabled, (event, targets) in zip(events, self.program.code):
      op, *args = event
      if op in TESTS:
        left, right = args
        if op == "symbols":
          a, b = self.tapes[left].focus, self.tapes[right].focus
          circuit.require(conjunction(neg(a.eq("_")), neg(b.eq("_"))), enabled)
          decision = a.equal(b)
        else:
          decision = (self.bank.equal if op == "equal" else self.bank.less)(left, right)
        edge(targets[0], conjunction(enabled, neg(decision)))
        edge(targets[1], conjunction(enabled, decision))
      elif op == "halt":
        halts.append(enabled)
      else:
        edge(targets[0], enabled)
        if op == "move":
          if len(args[0]) != 1 or abs(args[0][0][1]) != 1:
            raise ValueError("unit head instruction required")
          moves[args[0][0]].append(enabled)
        elif op == "copy":
          pair = tuple(args)
          copies[pair] = disjunction(copies.get(pair, FALSE), enabled)
        elif op == "border":
          borders.append(enabled)
        else:
          raise ValueError("unsupported GS instruction")

    # Input loading, rewind, and execution are disjoint modes. Therefore at
    # most one of these head movements allocates cells in this physical step.
    moves["OriginalEnd", 1].append(load)
    moves["Origin", -1].append(conjunction(rewind, neg(at_origin)))
    self.tapes["OriginalEnd"].write(Value({s: char.eq(s) for s in "ab#"}), load)
    if self.compact:
      self.bank.execute(events)
      self.bank.load_one(load)
    for (head, direction), conditions in moves.items():
      enabled = disjunction(*conditions)
      if enabled == FALSE:
        continue
      if not self.compact:
        self.bank.move(head, direction, enabled)
      if head in self.tapes:
        self.tapes[head].move(direction, enabled, slot=0)
    for (target, source), enabled in copies.items():
      self.copy(target, source, enabled)
    self.copy("Origin", "OriginalEnd", finish_load)
    for target in HEADS:
      if target not in ("Origin", "OriginalEnd"):
        self.copy(target, "Origin", begin)
    if self.compact:
      self.bank.initialize({("Origin", "OriginalEnd"): (self.bank.length, -1)}, begin)
    halt = disjunction(*halts)
    self.border = disjunction(*borders)
    self.found = disjunction(self.found, self.border)
    self.pc = Value.select(conjunction(active, neg(halt)), Value(next_pc), self.pc)
    self.mode = Value.select(finish_load, Value.constant("rewind"), self.mode)
    self.mode = Value.select(begin, Value.constant("run"), self.mode)
    self.mode = Value.select(halt, Value.constant("done"), self.mode)

  def finalize(self):
    self.bank.finalize()
    for tape in self.tapes.values():
      tape.finalize()
    self.circuit.put(self.pc_key, self.pc)
    self.circuit.put(self.mode_key, self.mode)
    self.circuit.put(self.found_key, Value.select(self.found, Value.constant(True), Value.constant(False)))


def fixture(program=None, compact=True):
  circuit = Circuit("ab#!.")
  worker = GSHeads(circuit, program, compact=compact)
  worker.tick()
  worker.finalize()
  circuit.put("gs.border", Value.select(worker.border, Value.constant(True), Value.constant(False)),
              (False, True), False)
  machine = circuit.machine(conjunction(worker.mode.eq("done"), worker.found))
  return circuit, worker, machine

File: scala/pal/src/main/scala/pal/ScaffoldGsHeads.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable
import Expr.FALSE
import Ref.PREVIOUS
import ScaffoldCircuit.{conjunction, disjunction, neg}

/** Finite GS head table lowered to one-instruction scaffold transitions.
  * Loading and rewinding are actual transitions; the fixture is not PAL.
  */
final class ScaffoldGsHeads(val circuit: Circuit, source: Option[Program] = None,
                    prefix: String = "gs", val compact: Boolean = true) {
  val program: Program = GsHeads.unitMoves(source.getOrElse(GsHeads.compileController()))
  val bank: Option[LiveDistances] = if (compact) {
    Some(new LiveDistances(circuit, program, GsHeads.HEADS, prefix + ".distance"))
  } else { None }
  val fullBank: Option[HeadDistances] = if (compact) { None } else {
    Some(new HeadDistances(circuit, GsHeads.HEADS, prefix = prefix + ".distance", exclusiveMoves = true))
  }
  val pool = new StackPool(circuit, Vector("cells" -> 1), Vector('_', 'a', 'b', '#'))
  val tapes: VectorMap[String, Tape] = VectorMap.from(GsHeads.HEADS.filterNot(GsHeads.BLIND).map { head =>
    head -> new Tape(circuit, prefix + "." + head, pool.payload, pool = Some(pool))
  })
  val pcKey: String = prefix + ".pc"
  val modeKey: String = prefix + ".mode"
  var pc: Value[Int] = circuit.get(PREVIOUS, pcKey, program.code.indices.toVector, program.start)
  var mode: Value[String] = circuit.get(PREVIOUS, modeKey, Vector("load", "rewind", "run", "done"), "load")
  val foundKey: String = prefix + ".found"
  var found: Expr = circuit.get(PREVIOUS, foundKey, Vector(false, true), false).eqTo(true)
  var border: Expr = FALSE

  private def equal(left: String, right: String): Expr = {
    bank.fold(fullBank.get.equal(left, right))(_.equal(left, right))
  }
  private def less(left: String, right: String): Expr = {
    bank.fold(fullBank.get.less(left, right))(_.less(left, right))
  }

  def copy(target: String, source: String, enabled: Expr): Unit = {
    fullBank.foreach(_.copy(target, source, enabled))
    if (tapes.contains(target)) {
      if (!tapes.contains(source)) { throw new IllegalArgumentException("a data head cannot copy a blind coordinate") }
      val a = tapes(target)
      val b = tapes(source)
      a.left.copyFrom(b.left, enabled)
      a.right.copyFrom(b.right, enabled)
      a.write(b.focus, enabled)
    }
  }

  def tick(): Unit = {
    val char = circuit.input()
    val load = conjunction(mode.eqTo("load"), disjunction("ab#".map(char.eqTo)*))
    val finishLoad = conjunction(mode.eqTo("load"), char.eqTo('!'))
    val rewind = conjunction(mode.eqTo("rewind"), char.eqTo('.'))
    val atOrigin = if (compact) { tapes("Origin").left.empty() } else { equal("Origin", "First") }
    val begin = conjunction(rewind, atOrigin)
    val active = conjunction(mode.eqTo("run"), char.eqTo('.'))
    val moves = mutable.LinkedHashMap.from(for (head <- GsHeads.HEADS; d <- Vector(-1, 1)) yield {
      (head, d) -> mutable.ArrayBuffer.empty[Expr]
    })
    val copies = mutable.LinkedHashMap.empty[(String, String), Expr]
    val nextPc = mutable.LinkedHashMap.empty[Int, Expr]
    val halts = mutable.ArrayBuffer.empty[Expr]
    val borders = mutable.ArrayBuffer.empty[Expr]
    def edge(destination: Int, enabled: Expr): Unit = {
      nextPc(destination) = disjunction(nextPc.getOrElse(destination, FALSE), enabled)
    }
    val events = program.code.indices.map(state => conjunction(active, pc.eqTo(state))).toVector
    for ((enabled, Row(event, targets)) <- events.zip(program.code)) {
      def branch(decision: Expr): Unit = {
        edge(targets(0), conjunction(enabled, neg(decision)))
        edge(targets(1), conjunction(enabled, decision))
      }
      event match {
        case Event.Symbols(left, right) =>
          val a = tapes(left).focus
          val b = tapes(right).focus
          circuit.require(conjunction(neg(a.eqTo('_')), neg(b.eqTo('_'))), enabled)
          branch(a.equal(b))
        case Event.Equal(left, right) => branch(equal(left, right))
        case Event.Less(left, right) => branch(less(left, right))
        case Event.Halt => halts += enabled
        case other =>
          edge(targets(0), enabled)
          other match {
            case Event.Move(batch) =>
              if (batch.size != 1 || math.abs(batch.head.delta) != 1) {
                throw new IllegalArgumentException("unit head instruction required")
              }
              moves((batch.head.head, batch.head.delta)) += enabled
            case Event.Copy(target, source) =>
              val pair = (target, source)
              copies(pair) = disjunction(copies.getOrElse(pair, FALSE), enabled)
            case Event.Border(_) => borders += enabled
            case _ => throw new IllegalArgumentException("unsupported GS instruction")
          }
      }
    }
    moves(("OriginalEnd", 1)) += load
    moves(("Origin", -1)) += conjunction(rewind, neg(atOrigin))
    tapes("OriginalEnd").write(Value("ab#".map(s => s -> char.eqTo(s))), load)
    bank.foreach { distances => distances.execute(events); distances.loadOne(load) }
    for (((head, direction), conditions) <- moves) {
      val enabled = disjunction(conditions.toSeq*)
      if (enabled != FALSE) {
        fullBank.foreach(_.move(head, direction, enabled))
        tapes.get(head).foreach(_.move(direction, enabled, Some(0)))
      }
    }
    for (((target, source), enabled) <- copies) { copy(target, source, enabled) }
    copy("Origin", "OriginalEnd", finishLoad)
    for (target <- GsHeads.HEADS if target != "Origin" && target != "OriginalEnd") { copy(target, "Origin", begin) }
    bank.foreach(distances => distances.initialize(Map(("Origin", "OriginalEnd") -> Some((distances.length, -1))), begin))
    val halt = disjunction(halts.toSeq*)
    border = disjunction(borders.toSeq*)
    found = disjunction(found, border)
    pc = Value.select(conjunction(active, neg(halt)), Value(nextPc), pc)
    mode = Value.select(finishLoad, Value.constant("rewind"), mode)
    mode = Value.select(begin, Value.constant("run"), mode)
    mode = Value.select(halt, Value.constant("done"), mode)
  }

  def commit(): Unit = {
    bank.foreach(_.commit())
    fullBank.foreach(_.commit())
    tapes.values.foreach(_.commit())
    circuit.put(pcKey, pc)
    circuit.put(modeKey, mode)
    circuit.put(foundKey, Value.select(found, Value.constant(true), Value.constant(false)))
  }
}

object ScaffoldGsHeads {
  def fixture(program: Option[Program] = None, compact: Boolean = true): (Circuit, ScaffoldGsHeads, Scaffold) = {
    val circuit = new Circuit("ab#!.")
    val worker = new ScaffoldGsHeads(circuit, program, compact = compact)
    worker.tick()
    worker.commit()
    circuit.put("gs.border", Value.select(worker.border, Value.constant(true), Value.constant(false)), Vector(false, true), false)
    (circuit, worker, circuit.machine(conjunction(worker.mode.eqTo("done"), worker.found)))
  }
}
